i3lock.c

```diff
@@ -645,6 +645,113 @@ void handle_screen_resize(void) {
     redraw_screen();
 }
 
+static int read_raw_image_native(uint32_t *dest, FILE *src, int width, int height, int stride) {
+    int count = 0;
+    for (int y = 0; y < height; ++y) {
+        size_t n = fread(&dest[y * stride], 1, width * 4, src);
+        count += n;
+        if (n < (size_t)(width * 4))
+            break;
+    }
+
+    return count;
+}
+
+static int read_raw_image_rgb(uint32_t *dest, FILE *src, int width, int height, int stride) {
+    unsigned char *buf = malloc(width * 3);
+    if (buf == NULL)
+        return -1;
+
+    int count = 0;
+    for (int y = 0; y < height; ++y) {
+        size_t n = fread(buf, 1, width * 3, src);
+        count += n;
+        if (n < (size_t)(width * 3))
+            break;
+
+        for (int x = 0; x < width; ++x) {
+            int idx = x * 3;
+            dest[y * stride + x] = 0 |
+                                   (buf[idx + 0] << 16) |
+                                   (buf[idx + 1] << 8) |
+                                   (buf[idx + 2]);
+        }
+    }
+
+    free(buf);
+    return count;
+}
+
+static cairo_surface_t *read_raw_image(const char *image_path, const char *image_raw_format) {
+    cairo_surface_t *img;
+
+    /* Parse format as <width>x<height>[:<format>] */
+    int w = -1, h = -1;
+    char format[10];
```

Comment:
It's 10 because the format option has to be hard coded as part of the format string for scanf, as it doesn't support `%.*s` like printf. We could do `char format[strlen("native") + 1]`, but scanf would then have to be `"%ix%i:%6s"`. I don't know that there's a nice way to fix this.

Reply:
What’s wrong with the sscanf format "%ix%i:%6s"? Seems fine to me.

Reply:
It's just that when using `char format[strlen("native")+1]`, it sort of looks like we can change the max length of a format by changing "native" to some other word, and not notice the significance of the hard-coded `%6s` on the next line. When the number is hard-coded, I think one would read the surrounding code more carefully before changing it.

In any case, I'm not extremely fond of either solution, so if you think changing it to `char format[strlen("native") + 1]` (or maybe rather `char format[sizeof("native")]` to make it a statically sized array) is better, I'm fine with that.

Reply:
Ah, now I understand your concern. How about this, then?
```c
  char format[] = "native";
  char *fmtstring = NULL;
  if (asprintf(&fmtstring, "%%%lds", sizeof(format)-1) == -1) {
    errx(…)
  }
  sscanf(…, fmtstring, …);
  free(fmtstring);  
```

Reply:
I pushed a commit which implements that now. However, note that asprintf is a GNU thing, not standard C.

I would personally have just gone with using a bigger than necessary fixed sized array (maybe a nice round size like 16 rather than my arbitrary 10) instead of relying on non-standard APIs which allocate memory for such a small thing, but I also completely understand the desire to not have sizes hard-coded like that.